File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
from collections import deque
input = sys.stdin.readline

N, M, V = map(int, input().split())
adjL = [[] for _ in range(N + 1)]
for _ in range(M):
    v, w = map(int, input().split())
    adjL[v].append(w)
    adjL[w].append(v)


def dfs(start):
    result = []
    stack = [start]
    visited = [0 for _ in range(N + 1)]

    while stack:
        v = stack.pop()
        if visited[v]:
            continue

        visited[v] = 1
        result.append(v)
        for w in sorted(adjL[v], reverse=True):    # 번호 큰 것부터 스택에 넣기
            if not visited[w]:
                stack.append(w)
                
    return result


def bfs(start):
    result = []
    q = deque([start])
    visited = [0 for _ in range(N + 1)]
    visited[start] = 1

    while q:
        v = q.popleft()
        result.append(v)

        for w in sorted(adjL[v]):
            if not visited[w]:
                visited[w] = 1
                q.append(w)
                
    return result

print(*dfs(V))
print(*bfs(V))

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.cc
#include <iostream>
#include <vector>
#include <stack>
#include <queue>
#include <algorithm>

using namespace std;

void dfs(vector<vector<int>>& graph, vector<int>& visited, int v) {
    stack<int> stk; stk.push(v);
    while (!stk.empty()) {
        v = stk.top(); stk.pop();
        if (!visited[v]) {
            visited[v] = true;
            cout << v << " ";

            for (auto it = graph[v].rbegin(); it != graph[v].rend(); it++) {
                stk.push(*it);
            }
        }
    }
}

void bfs(vector<vector<int>>& graph, vector<int>& visited, int v) {
    queue<int> q; q.push(v);
    visited[v] = true;
    cout << v << " ";
    
    while (!q.empty()) {
        v = q.front(); q.pop();

        for (const auto& w: graph[v]) {
            if (!visited[w]) {
                visited[w] = true;
                q.push(w);
                cout << w << " ";
            }
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int N, M, V;
    cin >> N >> M >> V;

    // 인접행렬
    vector<vector<int>> adjL(N + 1);
    int v, w;
    while (M--) {
        cin >> v >> w;
        adjL[v].push_back(w);
        adjL[w].push_back(v);
    }
    for (int i = 1; i < N + 1; i++) {
        sort(adjL[i].begin(), adjL[i].end());  // 오름차순 정렬
    }

    vector<int> visited(N + 1, false);
    dfs(adjL, visited, V);
    cout << "\n";
    
    fill(visited.begin(), visited.end(), false);
    bfs(adjL, visited, V);
    return 0;
}